streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -368,6 +370,7 @@ public void setUncaughtExceptionHandler(final Thread.UncaughtExceptionHandler eh
      * @param queryableStoreType    accept only stores that are accepted by {@link QueryableStoreType#accepts(StateStore)}
      * @param <T>                   return type
      * @return  A facade wrapping the {@link org.apache.kafka.streams.processor.StateStore} instances
+     * @throws org.apache.kafka.streams.errors.InvalidStateStoreException if the streams are (re-)initializing
```

Comment:
A few correlated comments:

1. Seems we are doing redundant check of `allStores.isEmpty()` in `QueryableStoreProvider` as the same check is also conducted in the returned `WrappingStoreProvider`.

2. And in `WrappingStoreProvider`, we could just return the empty list if `allStores.isEmpty()`; and similarly in `StreamThreadStateStoreProvider`, simply remove the `if (!store.isOpen())` check and still put it into the list since the check will be conducted later on the `get` / etc calls as well.

3. Finally in the `CompositeReadOnlyKeyValueStore` and `CompositeReadOnlyWindowStore`, check if `stores.isEmpty()` and also catch-rethrow `InvalidStateStoreException` thrown from `store.get/range/etc` calls to re-write the error message from

```
Store X is currently closed.
```

to

```
State store is not available anymore and may have been migrated to another instance; please re-discover its location from the state metadata.
```

And if `stores.isEmpty` also throw the `InvalidStateStoreException` with the same error message.

My understanding is that we can be ensured this is the case since in `onPartitionsAssigned`, we always update the state metadata after all the tasks have been added (i.e. stores being constructed and opened), is that right?